Multiple projects

dododedodonl opened the pull request:
This pull requests adds code to support multiple firebase projects in laravel via a new facade.

Since there is a breaking change in the configuration file, a major version bump is needed. Also, all existing facades are deprecated in favor of a single new facade to support multiple projects. They are however included for an easier upgrade path.

jeromegamez commented on src/FirebaseProject.php:
```diff
@@ -0,0 +1,106 @@
+
+class FirebaseProject
+{
+    /** @var \Kreait\Firebase\Factory $factory */
```
I usually don't include FQNs in PHPDocs, so this "risks" being changed to `Kreait\Firebase\Factory` (due to the existing import) or to `Factory` with an additional `use` statement somewhen in the future 🙈  (similarly elsewhere)

jeromegamez commented on src/FirebaseProject.php:
```diff
@@ -0,0 +1,106 @@
+    protected $config;
+
+    /** @var \Kreait\Firebase\Auth|null $auth */
+    protected $auth = null;
```
```suggestion
    protected $auth;
```

`null` is already the default value - it of course doesn't matter in the functionality, but while I usually like being explicit, class properties that are null by default are one exception 🙈 (even if you'll certainly find occurrences where I don't follow my own rules 😂)

dododedodonl replied:
Forgot to remove it when I removed the property types :)

jeromegamez commented on src/FirebaseProject.php:
```diff
@@ -0,0 +1,106 @@
+
+    public function auth(): Firebase\Auth
+    {
+        if (! $this->auth) {
```
```suggestion
        if (!$this->auth) {
```
Again, this is just style, but this is something I would otherwise change in a future commit, or as soon as I introduce PHP CS Fixer (with similar rules as in `kreait/firebase-php`)

jeromegamez commented on src/FirebaseProjectManager.php:
```diff
@@ -0,0 +1,122 @@
+
+class FirebaseProjectManager
+{
+    /** @var \Illuminate\Contracts\Foundation\Application */
```
Could you check if this equally works with `Illuminate\Contracts\Container\Container`? This would probably mean that getting the config would have to be implemented differently, but I remember changing `Application` to `Container` in the past for better Lumen support.

dododedodonl replied:
I'll have a look into it. I changed it to copy the `resolveCredentials` method to `FirebaseProjectManager`, as `basePath` does not exist on `Container`.

jeromegamez commented on src/ServiceProvider.php:
```diff
@@ -5,6 +5,7 @@
 namespace Kreait\Laravel\Firebase;
 
 use Illuminate\Contracts\Container\Container;
+use Illuminate\Contracts\Foundation\Application;
```
Same as in `FirebaseProjectManager` - I used `Container` in other parts of the app because of Lumen and if it's possible, we should stick with that

jeromegamez reviewed (changes requested):
That's a great PR, thank you so much! Most of my change requests are not concerning the new behavior, just my personal preference as a maintainer, I hope you can see past that 😅 

I'll try to test this for real soon, but this looks really, really good!

jeromegamez commented:
Could you please also remove support for Laravel/Illuminate 5.8 in the composer.json and GitHub actions? I think `orchestra/testbench ^3.8.6` can then go as well.

dododedodonl commented on README.md:
```diff
@@ -10,12 +10,17 @@ A Laravel package for the [Firebase PHP Admin SDK](https://github.com/kreait/fir
-* [Documentation](https://firebase-php.readthedocs.io/)
-- [Support](#support)
-- [License](#license)
+- [Firebase for Laravel](#firebase-for-laravel)
```
I had a plugin enabled that auto generated a table of contents, did not notice it

dododedodonl commented on src/FirebaseProjectManager.php:
```diff
@@ -0,0 +1,122 @@
+    protected $app;
+
+    /** @var FirebaseProject[] */
+    private $projects = [];
```
Missed that one, thanks for finding it :)

dododedodonl commented:
I think I incorporated all your suggestions :smiley: 
 
* Added docblock suggestion for facades
* Added php-cs-fixer (copied config from kreait/firebase-php, commented out 2 rules)
* Moved upgrade guide to UPGRADE.md
* Removed Laravel and Lumen 5.8 support, removed orchestra/testbench:^3.6.8, lowered minimums to 4.0 and 5.0
* Changed `Illuminate\Contracts\Foundation\Application` to `Illuminate\Contracts\Container\Container`
* Other small fixes

jeromegamez commented:
Sorry for the late response (again) - I just tested it locally and it works just fine, and using the new Facade is so much more enjoyable!

I will merge this now, and after integrating the other open PR and making some minor changes, I'll release this as a new version! This will be a nice new major release, thank you so much! 🌺
